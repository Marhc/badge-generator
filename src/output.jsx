// @flow

import React, { Component } from 'react';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import Radio from '@material-ui/core/Radio';
import RadioGroup from '@material-ui/core/RadioGroup';
import Typography from '@material-ui/core/Typography';
import { withStyles } from '@material-ui/core/styles';
import { type Service, type Format, formats } from './config';
import { type InputEvent } from './util';

type OutputProps = {|
  services: Service[],
  repository: string,
  badgeStyle: string,
  classes: {
    subtitle: string,
    badges: string,
    badge: string,
    formats: string,
    radioGroup: string,
    code: string,
  },
|};

type OutputState = {|
  formatIndex: number,
|};

const styles = ({ spacing, typography }) => ({
  subtitle: {
    marginTop: 2 * spacing.unit,
  },
  badges: {
    lineHeight: '20px',
  },
  badge: {
    marginRight: spacing.unit / 2,
  },
  formats: {
    marginTop: spacing.unit,
  },
  radioGroup: {
    flexDirection: 'row',
  },
  code: {
    fontFamily: 'Roboto Mono, monospace',
    fontSize: typography.fontSize - 3,
    lineHeight: 1.8,
    overflowX: 'auto',
    paddingBottom: spacing.unit,
  },
});

function formatUrl(url: string, repository: string, branch: string) {
  return url.replace('{repository}', repository).replace('{branch}', branch);
}

function nameToKey(name: string) {
  return name.replace(/ /g, '-').toLowerCase();
}

export class Output extends Component<OutputProps, OutputState> {
  state = {
    formatIndex: formats.findIndex(f => f.default) || 0,
  };

  get format(): Format {
    const { formatIndex } = this.state;
    return formats[formatIndex];
  }

  get formattedServices(): Service[] {
    const { services, repository, badgeStyle } = this.props;
    const branch = 'master';
    return services.map(({ url, imageUrl, ...props }) => ({
      url: formatUrl(url, repository, branch),
      imageUrl: `${formatUrl(
        imageUrl,
        repository,
        branch,
      )}?style=${badgeStyle}`,
      ...props,
    }));
  }

  handleFormatChange = (event: InputEvent, value: string) => {
    this.setState({ formatIndex: +value });
  };

  render() {
    const { classes } = this.props;
    const { formatIndex } = this.state;
    return (
      <>
        <Typography
          variant="h6"
          className={classes.subtitle}
          headlineMapping={{ h6: 'h2' }}
        >
          Preview
        </Typography>
        <p className={classes.badges}>
          {this.formattedServices.map(({ name, title, url, imageUrl }) => (
            <a
              key={`image-${nameToKey(name)}`}
              href={url}
              className={classes.badge}
              title={title}
            >
              <img src={imageUrl} alt={title} />
            </a>
          ))}
        </p>
        <Typography
          variant="h6"
          className={classes.subtitle}
          headlineMapping={{ h6: 'h2' }}
        >
          Source code
        </Typography>
        <form noValidate autoComplete="off" className={classes.formats}>
          <RadioGroup
            aria-label="Format"
            value={`${formatIndex}`}
            onChange={this.handleFormatChange}
            className={classes.radioGroup}
          >
            {formats.map(({ name }, index) => (
              <FormControlLabel
                key={`format-${nameToKey(name)}`}
                value={`${index}`}
                control={<Radio />}
                label={name}
              />
            ))}
          </RadioGroup>
        </form>
        <pre className={classes.code}>
          {this.formattedServices.map(
            ({ title, url, imageUrl }) =>
              `${this.format.template(title, url, imageUrl)}\n`,
          )}
        </pre>
      </>
    );
  }
}

export default withStyles(styles)(Output);
