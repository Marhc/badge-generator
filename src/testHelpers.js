// @flow

const toString = value => Object.prototype.toString.apply(value);

export function findByType(testInstance: ReactTestInstance, Type: Object) {
  const { instance } = testInstance.findByType(Type);
  if (!(instance instanceof Type)) {
    throw new TypeError(
      `Expected app to be an instance of ${Type.name}, got ${toString(
        instance,
      )}`,
    );
  }
  return instance;
}
