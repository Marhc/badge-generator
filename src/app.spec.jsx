// @flow

import React from 'react';
import ReactDOM from 'react-dom';
import * as TestRenderer from 'react-test-renderer';
import ShallowTestRenderer from 'react-test-renderer/shallow';
import AppWithStyle, { App } from './app';
import { Input } from './input';
import { Output } from './output';
import { services } from './config';
import { findByType } from './testHelpers';

describe('App', () => {
  describe('when repository is not set and services are not checked', () => {
    it('renders without crashing', () => {
      const div = document.createElement('div');
      ReactDOM.render(<AppWithStyle />, div);
      ReactDOM.unmountComponentAtNode(div);
    });

    it('renders its components', () => {
      const renderer = new ShallowTestRenderer();
      renderer.render(<App classes={{ container: '', paper: '' }} />);
      expect(renderer.getRenderOutput()).toMatchSnapshot();
    });

    it('renders <Input> component', () => {
      const renderer = TestRenderer.create(<AppWithStyle />);
      expect(renderer.root.findByType(Input).props).toMatchSnapshot();
    });

    it('does not render <Output> component', () => {
      const renderer = TestRenderer.create(<AppWithStyle />);
      expect(renderer.root.findAllByType(Output).length).toBe(0);
    });
  });

  describe('when repository is set and services are checked', () => {
    const state = {
      repository: 'REPOSITORY',
      serviceSelection: services.map(() => true),
    };

    it('updates <Input> component', () => {
      const renderer = TestRenderer.create(<AppWithStyle />);
      findByType(renderer.root, App).setState(state);
      expect(renderer.root.findByType(Input).props).toMatchSnapshot();
    });

    it('renders <Output> component', () => {
      const renderer = TestRenderer.create(<AppWithStyle />);
      findByType(renderer.root, App).setState(state);
      expect(renderer.root.findByType(Output).props).toMatchSnapshot();
    });
  });

  describe('#handleRepositoryChange()', () => {
    it('is a function', () => {
      const renderer = TestRenderer.create(<AppWithStyle />);
      const app = findByType(renderer.root, App);
      expect(app.handleRepositoryChange).toBeFunction();
    });

    it('is passed to <Input> component as is', () => {
      const renderer = TestRenderer.create(<AppWithStyle />);
      const app = findByType(renderer.root, App);
      expect(renderer.root.findByType(Input).props.handleRepositoryChange).toBe(
        app.handleRepositoryChange,
      );
    });

    it("updates App 'repository' state", () => {
      const renderer = TestRenderer.create(<AppWithStyle />);
      const app = findByType(renderer.root, App);
      expect(app.state.repository).not.toBe('REPOSITORY');

      const target = { value: 'REPOSITORY' };
      app.handleRepositoryChange({ target });
      expect(app.state.repository).toBe('REPOSITORY');
    });
  });

  describe('#handleServiceToggle()', () => {
    it('is a function', () => {
      const renderer = TestRenderer.create(<AppWithStyle />);
      const app = findByType(renderer.root, App);
      expect(app.handleServiceToggle).toBeFunction();
    });

    it('is passed to <Input> component as is', () => {
      const renderer = TestRenderer.create(<AppWithStyle />);
      const app = findByType(renderer.root, App);
      expect(renderer.root.findByType(Input).props.handleServiceToggle).toBe(
        app.handleServiceToggle,
      );
    });

    it("returns a function that updates App 'serviceSelection' state", () => {
      const renderer = TestRenderer.create(<AppWithStyle />);
      const app = findByType(renderer.root, App);
      expect(app.state.serviceSelection[0]).not.toBe(true);

      const handleFirstServiceToggle = app.handleServiceToggle(0);
      expect(handleFirstServiceToggle).toBeFunction();

      const target = { checked: true };
      handleFirstServiceToggle({ target });
      expect(app.state.serviceSelection[0]).toBe(true);
    });
  });

  describe('#handleStyleChange()', () => {
    it('is a function', () => {
      const renderer = TestRenderer.create(<AppWithStyle />);
      const app = findByType(renderer.root, App);
      expect(app.handleStyleChange).toBeFunction();
    });

    it('is passed to <Input> component as is', () => {
      const renderer = TestRenderer.create(<AppWithStyle />);
      const app = findByType(renderer.root, App);
      expect(renderer.root.findByType(Input).props.handleStyleChange).toBe(
        app.handleStyleChange,
      );
    });

    it("updates App 'style' state", () => {
      const renderer = TestRenderer.create(<AppWithStyle />);
      const app = findByType(renderer.root, App);
      expect(app.state.style).not.toBe('STYLE');

      app.handleStyleChange({}, 'STYLE');
      expect(app.state.style).toBe('STYLE');
    });
  });
});
