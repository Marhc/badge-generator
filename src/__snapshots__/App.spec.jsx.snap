// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App when repository is not set and services are not checked renders <Input> component 1`] = `
Object {
  "badgeStyle": "default",
  "classes": Object {
    "gridItem": "Input-gridItem-337",
  },
  "handleRepositoryChange": [Function],
  "handleServiceToggle": [Function],
  "handleStyleChange": [Function],
  "repository": "",
  "serviceSelection": Array [
    false,
    false,
    false,
    false,
    false,
    false,
    false,
    false,
    false,
    false,
    false,
    false,
    false,
    false,
    false,
    false,
    false,
    false,
    false,
    false,
    false,
    false,
    false,
    false,
    false,
    false,
    false,
    false,
  ],
}
`;

exports[`App when repository is not set and services are not checked renders its components 1`] = `
<main
  className=""
>
  <WithStyles(CssBaseline) />
  <WithStyles(Typography)
    headlineMapping={
      Object {
        "h4": "h1",
      }
    }
    variant="h4"
  >
    Github Badge Generator
  </WithStyles(Typography)>
  <WithStyles(Paper)
    className=""
  >
    <WithStyles(Input)
      badgeStyle="default"
      handleRepositoryChange={[Function]}
      handleServiceToggle={[Function]}
      handleStyleChange={[Function]}
      repository=""
      serviceSelection={
        Array [
          false,
          false,
          false,
          false,
          false,
          false,
          false,
          false,
          false,
          false,
          false,
          false,
          false,
          false,
          false,
          false,
          false,
          false,
          false,
          false,
          false,
          false,
          false,
          false,
          false,
          false,
          false,
          false,
        ]
      }
    />
  </WithStyles(Paper)>
  
</main>
`;

exports[`App when repository is set and services are checked renders <Output> component 1`] = `
Object {
  "badgeStyle": "default",
  "classes": Object {
    "badge": "Output-badge-545",
    "badges": "Output-badges-544",
    "code": "Output-code-548",
    "formats": "Output-formats-546",
    "radioGroup": "Output-radioGroup-547",
    "subtitle": "Output-subtitle-543",
  },
  "repository": "REPOSITORY",
  "services": Array [
    Object {
      "imageUrl": "https://img.shields.io/github/forks/{repository}.svg",
      "name": "GitHub forks",
      "title": "Build Status",
      "url": "https://github.com/{repository}",
    },
    Object {
      "imageUrl": "https://img.shields.io/github/stars/{repository}.svg",
      "name": "GitHub stars",
      "title": "Build Status",
      "url": "https://github.com/{repository}",
    },
    Object {
      "imageUrl": "https://img.shields.io/github/license/{repository}.svg",
      "name": "GitHub license",
      "title": "License",
      "url": "https://github.com/{repository}",
    },
    Object {
      "imageUrl": "https://img.shields.io/travis/{repository}/{branch}.svg",
      "name": "Travis CI",
      "title": "Build Status",
      "url": "https://travis-ci.org/{repository}",
    },
    Object {
      "imageUrl": "https://img.shields.io/circleci/project/{repository}/{branch}.svg",
      "name": "CircleCI",
      "title": "Build Status",
      "url": "https://circleci.com/gh/{repository}",
    },
    Object {
      "imageUrl": "https://circleci.com/gh/{repository}.svg?style=svg",
      "name": "CodeShip",
      "title": "Build Status",
      "url": "https://codeship.com/projects/{repository}",
    },
    Object {
      "imageUrl": "https://img.shields.io/codeclimate/maintainability/{repository}.svg",
      "name": "Code Climate (maintainability)",
      "title": "Maintainability",
      "url": "https://codeclimate.com/github/{repository}",
    },
    Object {
      "imageUrl": "https://img.shields.io/codeclimate/maintainability-percentage/{repository}.svg",
      "name": "Code Climate (% maintainability)",
      "title": "Maintainability",
      "url": "https://codeclimate.com/github/{repository}",
    },
    Object {
      "imageUrl": "https://img.shields.io/codeclimate/tech-debt/{repository}.svg",
      "name": "Code Climate (tech debt)",
      "title": "Technical Debt",
      "url": "https://codeclimate.com/github/{repository}",
    },
    Object {
      "imageUrl": "https://img.shields.io/codeclimate/coverage/github/{repository}.svg",
      "name": "Code Climate (code coverage)",
      "title": "Test Coverage",
      "url": "https://codeclimate.com/github/{repository}",
    },
    Object {
      "imageUrl": "http://img.shields.io/coveralls/{repository}/{branch}.svg",
      "name": "Coveralls",
      "title": "Test Coverage",
      "url": "https://coveralls.io/r/{repository}?branch={branch}",
    },
    Object {
      "imageUrl": "https://img.shields.io/codecov/c/github/{repository}/{branch}.svg",
      "name": "Codecov",
      "title": "Test Coverage",
      "url": "https://codecov.io/github/{repository}?branch={branch}",
    },
    Object {
      "imageUrl": "http://img.shields.io/gemnasium/{repository}.svg",
      "name": "Gemnasium",
      "title": "Dependency Status",
      "url": "https://gemnasium.com/{repository}",
    },
    Object {
      "imageUrl": "https://www.versioneye.com/user/projects/{repository}/badge.svg",
      "name": "VersionEye",
      "title": "Dependency Status",
      "url": "https://www.versioneye.com/user/projects/{repository}",
    },
    Object {
      "imageUrl": "https://img.shields.io/david/{repository}.svg",
      "name": "David DM",
      "title": "Dependency Status",
      "url": "https://david-dm.org/{repository}",
    },
    Object {
      "imageUrl": "https://img.shields.io/david/dev/{repository}.svg",
      "name": "David DM (dev)",
      "title": "devDependency Status",
      "url": "https://david-dm.org/{repository}#info=devDependencies",
    },
    Object {
      "imageUrl": "https://img.shields.io/david/peer/{repository}.svg",
      "name": "David DM (peer)",
      "title": "peerDependency Status",
      "url": "https://david-dm.org/{repository}#info=devDependencies",
    },
    Object {
      "imageUrl": "https://img.shields.io/packagist/v/{repository}.svg",
      "name": "Packagist",
      "title": "Latest Stable Version",
      "url": "https://packagist.org/packages/{repository}",
    },
    Object {
      "imageUrl": "https://img.shields.io/packagist/php-v/{repository}.svg",
      "name": "Packagist PHP version",
      "title": "PHP version",
      "url": "https://packagist.org/packages/{repository}",
    },
    Object {
      "imageUrl": "https://img.shields.io/packagist/l/{repository}.svg",
      "name": "Packagist License",
      "title": "License",
      "url": "https://packagist.org/packages/{repository}",
    },
    Object {
      "imageUrl": "https://img.shields.io/npm/v/{repository}.svg",
      "name": "NPM",
      "title": "Latest Stable Version",
      "url": "https://www.npmjs.com/package/{repository}",
    },
    Object {
      "imageUrl": "https://img.shields.io/npm/l/{repository}.svg",
      "name": "NPM License",
      "title": "License",
      "url": "https://www.npmjs.com/package/{repository}",
    },
    Object {
      "imageUrl": "https://img.shields.io/npm/dt/{repository}.svg",
      "name": "NPM Total Downloads",
      "title": "NPM Downloads",
      "url": "https://www.npmjs.com/package/{repository}",
    },
    Object {
      "imageUrl": "https://img.shields.io/npm/dm/{repository}.svg",
      "name": "NPM Downloads per month",
      "title": "NPM Downloads",
      "url": "https://www.npmjs.com/package/{repository}",
    },
    Object {
      "imageUrl": "https://img.shields.io/bower/v/{repository}.svg",
      "name": "Bower",
      "title": "Latest Stable Version",
      "url": "http://bower.io/search/?q={repository}",
    },
    Object {
      "imageUrl": "https://doc.esdoc.org/github.com/{repository}/badge.svg",
      "name": "ESDoc",
      "title": "API Documentation",
      "url": "https://doc.esdoc.org/github.com/{repository}/",
    },
    Object {
      "imageUrl": "https://badges.greenkeeper.io/{repository}.svg",
      "name": "Greenkeeper",
      "title": "Greenkeeper",
      "url": "https://github.com/{repository}/issues?q=label%3Agreenkeeper",
    },
    Object {
      "imageUrl": "http://melpa.org/packages/{repository}-badge.svg",
      "name": "Melpa",
      "title": "MELPA",
      "url": "http://melpa.org/#/{repository}/",
    },
  ],
}
`;

exports[`App when repository is set and services are checked updates <Input> component 1`] = `
Object {
  "badgeStyle": "default",
  "classes": Object {
    "gridItem": "Input-gridItem-337",
  },
  "handleRepositoryChange": [Function],
  "handleServiceToggle": [Function],
  "handleStyleChange": [Function],
  "repository": "REPOSITORY",
  "serviceSelection": Array [
    true,
    true,
    true,
    true,
    true,
    true,
    true,
    true,
    true,
    true,
    true,
    true,
    true,
    true,
    true,
    true,
    true,
    true,
    true,
    true,
    true,
    true,
    true,
    true,
    true,
    true,
    true,
    true,
  ],
}
`;
